runtime-sdk/modules/evm/src/precompile/erc20.rs

```diff
@@ -0,0 +1,556 @@
+const STATE_ALLOWANCES: &[u8] = &[0x01];
+
+/// Errors emitted by the ERC-20 implementation.
+#[derive(
+    thiserror::Error, Debug, oasis_runtime_sdk_macros::Error, oasis_runtime_sdk_macros::EvmError,
+)]
+pub enum Error {
+    #[error("forbidden by policy")]
+    #[sdk_error(code = 1)]
+    #[evm_error(signature = "Forbidden()")]
+    Forbidden,
+
+    /// EIP-6093 defined error.
+    #[error("invalid sender address {0}")]
+    #[sdk_error(code = 2)]
+    #[evm_error(signature = "ERC20InvalidSender(address)")]
+    ERC20InvalidSender(H160),
+
+    /// EIP-6093 defined error.
+    #[error("invalid receiver address {0}")]
+    #[sdk_error(code = 3)]
+    #[evm_error(signature = "ERC20InvalidReceiver(address)")]
+    ERC20InvalidReceiver(H160),
+
+    /// EIP-6093 defined error.
+    #[error("invalid approver address {0}")]
+    #[sdk_error(code = 4)]
+    #[evm_error(signature = "ERC20InvalidApprover(address)")]
+    ERC20InvalidApprover(H160),
+
+    /// EIP-6093 defined error.
+    #[error("invalid spender address {0}")]
+    #[sdk_error(code = 5)]
+    #[evm_error(signature = "ERC20InvalidSpender(address)")]
+    ERC20InvalidSpender(H160),
+
+    /// EIP-6093 defined error.
+    #[error("insufficient balance: {0}, {1} < {2}")]
+    #[sdk_error(code = 6)]
+    #[evm_error(signature = "ERC20InsufficientBalance(address,uint256,uint256)")]
+    ERC20InsufficientBalance(H160, u128, u128),
+
+    /// EIP-6093 defined error.
+    #[error("insufficient balance: {0}, {1} < {2}")]
+    #[sdk_error(code = 7)]
+    #[evm_error(signature = "ERC20InsufficientAllowance(address,uint256,uint256)")]
+    ERC20InsufficientAllowance(H160, u128, u128),
+
+    #[error("accounts: {0}")]
+    #[sdk_error(transparent)]
+    #[evm_error(signature = "Accounts(string)")]
+    Accounts(#[from] oasis_runtime_sdk::modules::accounts::Error),
+}
+
+impl Error {
+    fn annotate_accounts_error<T>(
+        err: AccountsError,
+        account: address::Address,
+        address: H160,
+        amount: &token::BaseUnits,
+    ) -> Self
+    where
+        T: AccountToken,
+    {
+        match err {
+            AccountsError::InsufficientBalance => {
+                match T::Accounts::get_balance(account, amount.denomination().clone()) {
+                    Ok(balance) => {
+                        Self::ERC20InsufficientBalance(address, balance, amount.amount())
+                    }
+                    Err(err) => err.into(),
+                }
+            }
+            _ => err.into(),
+        }
+    }
+}
+
+/// The gas costs for the various ERC-20 token functions.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct TokenOperationCosts {
+    name: u64,
+    symbol: u64,
+    decimals: u64,
+    total_supply: u64,
+    balace_of: u64,
+    transfer: u64,
+    transfer_from: u64,
+    approve: u64,
+    allowance: u64,
+    mint: u64,
+    burn: u64,
+}
+
+impl TokenOperationCosts {
+    pub const fn new() -> Self {
+        Self {
```

Comment:
Why do you need `new` since it looks the same as `Default::default`? I would implement a custom `Default` instead and use non-zero values.

Reply:
I didn't implement `Default` because it can't be `const`, but that's also the only reason. Would it be better to forego constness and change the cost const in the trait to a function?